### PR TITLE
Docs: Add builtins documentation for cairo vm

### PR DESCRIPTION
**Pull request by jimenezz22**
This PR is for the issue Builtins Documentation #326. Adding a complete documentation to understand how builtins works on cairo virtual machine and how bulitins perform complex operations at the cost of extra constraints when proving.

### CODE REVIEW

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:35]: I am not aware of addition/sub/mul/div builtins. Could you share your source for this?

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:59]: Memory in Cairo is write-once, so there is never actually any deallocation. All memory written stays like that forever.

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:55]: I am not aware that there is a builtin that performs this operations.

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:71]: I am not aware of this builtins as well, could you share their names with me?

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:85]: This is not the correct way of using the output builtin

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:82]: The way this parameters are declared is incorrect. In any case it would be:

```
func add(a: felt, b: felt) -> (felt) {
```

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:31]: This whole section is wrong. Neither Cairo Zero nor Cairo One use builtins in this way

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:155]: Builtins come with a cost, but what you describe in this section is completely unrelated as far as my knowledge goes. I personally don't find the knowledge in this chapter beneficial for understanding builtins.

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:29]: Although the initial analogy with Legos is good, I don't think explaining the whole meaning with analogies is helpful rather than confusing. Bultins should be explained as what they are: special segments of the Cairo VM that behave a little different. The bring efficiency to the table by doing certain operations that with Cario set of instructions is extremely expensive at the cost of extra complexity when proving the program.

**Review by rodrigo-pino — CHANGES_REQUESTED** [review]
To be honest, I liked the beginning but after that it all started feeling more and more as something generated by an IA and no honest effort put into it.

Some general feedback: 
1. Is ok to write just what you understood of builtins.
2. If you wanted to write more, is ok for you to ask us questions, we will gladly answer or point towards some documentation.
3. Is ok to make mistakes

On the other hand is not Ok to fill the PR with content just for the sake of filling it with content. Usually the content doesn't adds anything meaningful, at instances, is slightly correct but it will cause more confusion than anything. Just by trying to run some Cairo Code (or reading the docs) you would realize that there are things written here that don't make sense (which begs the question, how much effort was put into this PR).

Some specific feedback:
1. The introduction can stay as is
2. `What is builtins in Cairo` and `Builtins and Cairo memory` should be explained as what they are without any analogies.
3. `General types of builtins in Cairo VM`: The concept is good, but it has builtins I don't know about. Also, `sepcific builtins in Cairo Vm` could be merged with this section.
4. Finally, `How builtins perform ...` should explain the trade-off that brings builtins to the table which has nothing to do with what was written in that section.

**Comment by jimenezz22** [comment]
Hi! @rodrigo-pino
I'm working on the changes and will make a new commit as soon as possible. Sorry for the use of analogies, I used them because it is easier for me to understand these types of concepts through the use of analogies and I thought it would be good to include them in the documentation. I will make documentation cleaner, shorter and with more technical language.

**Comment by jimenezz22** [comment]
@rodrigo-pino I already made the commit with the requested changes

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:65]: The code in this section is still wrong and doesn't reflect the actual usages of these builtins. It would be good for readers to easily copy the code into a cairo file and execute it.

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:67]: Output is great 👍

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:84]: Let's make all the remaining builtins described (`pedersen`,  `range_check`, `ecdsa` & `bitwise`) to be explained just as `output` is described

**Review by rodrigo-pino — COMMENTED** [review]
Much better @jimenezz22 ! Thank you. Just some comments

**Comment by jimenezz22** [comment]
Hi! @rodrigo-pino 
I'll make a new commit as soon as possible

File: docs/docs/vm-fundamentals/builtins.md
Comment (jimenezz22) [reply docs/docs/vm-fundamentals/builtins.md:65]: @rodrigo-pino can you tell me what is exactly wrong in this code please? I was guided by a very similar example in the Cairo documentation and I can't find what is wrong

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [reply docs/docs/vm-fundamentals/builtins.md:84]: meaning it's not cairo code. I cannot run the code in this examples because it won't compile

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [reply docs/docs/vm-fundamentals/builtins.md:65]: Is not cairo code (except for output). I doubt you found this code in any documentation. Hope you can realize the obvious difference between the code the `output` builtin and the others

**Comment by rodrigo-pino** [comment]
@jimenezz22 do you have any updates on this?

**Comment by jimenezz22** [comment]
@rodrigo-pino I was waiting for you to answer the questions I had about the comments. In the next few hours you will have the issue ready

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [reply docs/docs/vm-fundamentals/builtins.md:65]: Yep, I am not talking about the code above this comment, I am referring to the code examples of the subsection **Some specific builtins in Cairo VM**. As you'll notice, is very dissimilar to the code example above this comment which is the one I am not referring to. 

The only builtin of the builtins listed in this section which is an actual Cairo example is the `output` builtins. The remaining builtins, the ones defined below the `output` builtins, listed in order of appearance:
1. Pedersen
2. Range Check
3. ecdsa
4. bitwise

Are not valid examples of Cairo Code

Comment (jimenezz22) [reply docs/docs/vm-fundamentals/builtins.md:65]: Now I understand it much better, thank you

**Comment by jimenezz22** [comment]
@rodrigo-pino new commit ready

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:5]: You remove the lego part noooo 😱 

I liked that intro 😅

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:169]: > ... This pointer is __used__ to manage ...

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:271]: You can just say
`const ALL_ONES = -1`

**Review by rodrigo-pino — COMMENTED** [review]
Cool, just some nitpicks. Will approve once resolved

**Comment by jimenezz22** [comment]
@rodrigo-pino Changes made! Waiting for approval.

**Review by rodrigo-pino — APPROVED** [review]

File: docs/docs/vm-fundamentals/builtins.md
Comment (TAdev0) [inline docs/docs/vm-fundamentals/builtins.md:171]: which points to a variable of type `SignatureBuiltin`

or which is of type `SignatureBuiltin*`

File: docs/docs/vm-fundamentals/builtins.md
Comment (TAdev0) [inline docs/docs/vm-fundamentals/builtins.md:1]: why did you remove the sidebar position?

Also, please remove  `#326`

**Review by TAdev0 — APPROVED** [review]
left minor comments. Looks good overall,  a few improvments should be done but this can be done later in another PR

File: docs/docs/vm-fundamentals/builtins.md
Comment (rodrigo-pino) [inline docs/docs/vm-fundamentals/builtins.md:167]: The full sentence should be like this
```suggestion
A structure defines the memory layout for the signature builtin. It's employed by various functions in the common library, including the ecdsa builtin. For instance, in the verify_ecdsa_signature() function, there's an implicit argument of type `*SignatureBuiltin`, which aids in tracking the next available builtin instance.
```

**Pull request merged**
